packages/native/types/base.d.ts

```diff
@@ -0,0 +1,227 @@
+  | 'ToolbarAndroid'
+  | 'TouchableHighlight'
+  | 'TouchableNativeFeedback'
+  | 'TouchableOpacity'
+  | 'TouchableWithoutFeedback'
+  | 'View'
+  | 'ViewPagerAndroid'
+  | 'WebView'
+
+export type ReactNativeComponents = Pick<ReactNative, ReactNativeComponentNames>
+
+export type ReactNativeComponentProps<
+  ComponentName extends ReactNativeComponentNames
+> = ComponentPropsWithoutRef<ReactNativeComponents[ComponentName]>
+
+export type ReactNativeStyleType<Props> = Props extends {
+  style?: RN.StyleProp<infer StyleType>
+}
+  ? StyleType
+  : ReactNativeStyle
+
+export type ObjectInterpolation<StyleType = ReactNativeStyle> = StyleType
+
+export interface ArrayInterpolation<MergedProps, StyleType = ReactNativeStyle>
+  extends Array<Interpolation<MergedProps, StyleType>> {}
+
+export interface FunctionInterpolation<
+  MergedProps,
+  StyleType = ReactNativeStyle
+> {
+  (mergedProps: MergedProps): Interpolation<MergedProps, StyleType>
+}
+
+export type Interpolation<
+  MergedProps = unknown,
+  StyleType = ReactNativeStyle
+> =
+  | null
+  | undefined
+  | boolean
+  | number
+  | string
+  | ObjectInterpolation<StyleType>
+  | ArrayInterpolation<MergedProps, StyleType>
+  | FunctionInterpolation<MergedProps, StyleType>
+
+/** Same as StyledOptions but shouldForwardProp must be a type guard */
+export interface FilteringStyledOptions<
+  Props,
+  ForwardedProps extends keyof Props = keyof Props
+> {
+  shouldForwardProp?(propName: PropertyKey): propName is ForwardedProps
+}
+
+export interface StyledOptions<Props> {
+  shouldForwardProp?(propName: PropertyKey): boolean
+}
+
+export interface StyledWithComponent<ComponentProps extends {}> {
+  withComponent<
+    Component extends ComponentType<ComponentPropsWithoutRef<Component>>
+  >(
+    component: Component
+  ): StyledComponent<ComponentProps & ComponentPropsWithoutRef<Component>>
+  withComponent<ComponentName extends ReactNativeComponentNames>(
+    component: ReactNativeComponents[ComponentName]
+  ): StyledComponent<ComponentProps, ReactNativeComponentProps<ComponentName>>
+}
+
+/**
+ * @typeparam ComponentProps  Props which will be included when withComponent is called
+ * @typeparam SpecificComponentProps  Props which will *not* be included when withComponent is called
+ */
+export type StyledComponent<
+  ComponentProps extends {},
+  SpecificComponentProps extends {} = {}
+> = NamedExoticComponent<ComponentProps & SpecificComponentProps> &
+  StyledWithComponent<ComponentProps>
+
+/**
+ * @typeparam ComponentProps  Props which will be included when withComponent is called
+ * @typeparam SpecificComponentProps  Props which will *not* be included when withComponent is called
+ * @typeparam StyleProps  Params passed to styles but not exposed as React props. These are normally library provided props
+ */
+export interface CreateStyledComponent<
+  ComponentProps extends {},
+  SpecificComponentProps extends {} = {},
+  StyleProps extends {} = {},
+  StyleType = ReactNativeStyle
```

Comment:
I found that the `StyleType`, while properly inferred, was allowing an `ImageStyle` to be used as a `TextStyle` but not the other way around. Feels like a bug in the TypeScript inference logic because i couldn't see a relation between the two, but i couldn't get a better answer as to why (i left some comments in the test file).

Reply:
```ts
type True = RN.ImageStyle extends RN.TextStyle ? true : false
type False = RN.TextStyle extends RN.ImageStyle ? true : false
```

Seems like ImageStyle is a subtype of TextStyle

Reply:
haha interesting, i'm not going to dig further since it's already difficult to understand why that may be.

Reply:
Similarly on web input is a subtype of div. It's not like those types form such hierarchy, but rather that div is so basic, that it has no special props associated with it and everything that can be used on it can be used on input as well (+ input has some specific props for it). And because TS type system is structural rather than nominal this forms implicit supertype-subtype relationship.